File: app/src/ultimate/java/com/clockmods/widget/render/WidgetPendingIntentFactory.java
package com.clockmods.widget.render;
import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.content.*;
import android.net.Uri;
import com.clockmods.widget.config.WidgetConfigActivity;
import com.clockmods.widget.update.WidgetDataChangedReceiver;
import com.clockmods.ultimate.UltimateMainActivity;
import com.clockmods.ultimate.settings.UltimateSettingsActivity;
public final class WidgetPendingIntentFactory {
    public enum Action { CLOCK, CALENDAR, WEATHER, CONFIG, REFRESH }
    private WidgetPendingIntentFactory() { }
    public static int requestCode(int id,Action action) { return 31*id+action.ordinal(); }
    public static String identity(int id,Action action) { return "clockmods-widget://instance/"+id+"/"+action.name(); }
    public static PendingIntent create(Context c,int id,Action action) {
        Intent i;
        switch(action) {
            case CONFIG: i=new Intent(c,WidgetConfigActivity.class); break;
            case WEATHER: i=UltimateSettingsActivity.createSubpageIntent(c,"weather"); break;
            case REFRESH: i=new Intent(c,WidgetDataChangedReceiver.class).setAction(WidgetDataChangedReceiver.REFRESH); break;
            default: i=new Intent(c,UltimateMainActivity.class); break;
        }
        i.setData(Uri.parse(identity(id,action))).putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID,id);
        int flags=PendingIntent.FLAG_UPDATE_CURRENT|PendingIntent.FLAG_IMMUTABLE;
        return action==Action.REFRESH ? PendingIntent.getBroadcast(c,requestCode(id,action),i,flags)
            : PendingIntent.getActivity(c,requestCode(id,action),i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK),flags);
    }
    public static Action tap(String value) {
        switch(value) { case "open_config": return Action.CONFIG; case "open_weather": return Action.WEATHER; case "open_calendar": return Action.CALENDAR; default: return Action.CLOCK; }
    }
}

File: app/src/ultimate/java/com/clockmods/widget/render/WidgetPendingIntentFactory.kt
package com.clockmods.widget.render

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import com.clockmods.ultimate.ComposeMainActivity
import com.clockmods.ultimate.UltimateMainActivity
import com.clockmods.ultimate.settings.ComposeSettingsActivity
import com.clockmods.widget.config.WidgetConfigActivity
import com.clockmods.widget.update.WidgetDataChangedReceiver

object WidgetPendingIntentFactory {
    enum class Action { CLOCK, CALENDAR, WEATHER, CONFIG, REFRESH }

    @JvmStatic fun requestCode(id: Int, action: Action): Int = 31 * id + action.ordinal

    @JvmStatic fun identity(id: Int, action: Action): String =
        "clockmods-widget://instance/$id/${action.name}"

    @JvmStatic fun create(context: Context, id: Int, action: Action): PendingIntent {
        val intent = when (action) {
            Action.CONFIG -> Intent(context, WidgetConfigActivity::class.java)
            Action.WEATHER -> ComposeSettingsActivity.createSubpageIntent(context, "weather")
            Action.REFRESH -> Intent(context, WidgetDataChangedReceiver::class.java)
                .setAction(WidgetDataChangedReceiver.REFRESH)
            Action.CLOCK, Action.CALENDAR -> Intent(context, UltimateMainActivity::class.java)
                .putExtra(ComposeMainActivity.EXTRA_DESTINATION, destination(action))
        }.apply {
            data = Uri.parse(identity(id, action))
            putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, id)
        }
        val flags = PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        return if (action == Action.REFRESH) {
            PendingIntent.getBroadcast(context, requestCode(id, action), intent, flags)
        } else {
            PendingIntent.getActivity(context, requestCode(id, action), intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK), flags)
        }
    }

    @JvmStatic fun tap(value: String?): Action = when (value) {
        "open_config" -> Action.CONFIG
        "open_weather" -> Action.WEATHER
        "open_calendar" -> Action.CALENDAR
        else -> Action.CLOCK
    }

    @JvmStatic fun destination(action: Action): String? = when (action) {
        Action.CLOCK -> ComposeMainActivity.DESTINATION_CLOCK
        Action.CALENDAR -> ComposeMainActivity.DESTINATION_CALENDAR
        else -> null
    }
}
